### ISSUE TITLE
Faster media start up

### ISSUE THREAD

**Issue by mraavi** [google/shaka-player#2291]
Our application stream Live media.
Media Start up time is 4 to 5 sec.
I am looking for ways to reduce media start time up time in order to provide best user experience.
Following are the list of call during media start up.

<img width="333" alt="image" src="https://user-images.githubusercontent.com/6366393/70593454-318c9d80-1c03-11ea-8f51-4ff9a1d7a99f.png">

You can observe that the call in the list
MPD
init.mp4 for audio
1st audio segment
init.mp4 for video
1st video segment
get License call
2nd audio segment
3rd audio segment
2nd video segment
Media start only after downloading 2nd video segment. Segment duration is 6 sec.
Following are the abr and streaming configurations
```js
mPlayer.configure({
            abr: {
                defaultBandwidthEstimate: 1000000,
                switchInterval: 6
            },
            streaming: {
                bufferBehind: 6,
                bufferingGoal: 6
            }
        });
```

Following is the mpd,

```
<?xml version="1.0" encoding="utf-8"?>
<MPD xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:mobitv="urn:mobitv" xmlns="urn:mpeg:dash:schema:mpd:2011" xsi:schemaLocation="urn:mpeg:dash:schema:mpd:2011 DASH-MPD.xsd" profiles="urn:mpeg:dash:profile:isoff-live:2011,urn:com:dashif:dash264" maxSegmentDuration="PT6S" minBufferTime="PT2S" type="dynamic" timeShiftBufferDepth="PT20M" minimumUpdatePeriod="PT8760H" availabilityStartTime="1970-01-01T00:00:19Z" suggestedPresentationDelay="PT11S" id="5062" publishTime="2019-12-11T05:23:54Z">
   <ProgramInformation>
      <Title>Media Presentation Description by MobiTV.</Title>
   </ProgramInformation>
   <Period mobitv:publishTime="2019-12-11T05:23:54Z" id="precambrian" start="PT0S" mobitv:thumbnails="150_T;640_T">
      <AdaptationSet contentType="video" mimeType="video/mp4" mobitv:qualityLevel="SD" segmentAlignment="true" startWithSAP="1" par="16:9" minWidth="640" maxWidth="1280" minHeight="360" maxHeight="720" maxFrameRate="60/2">
         <Accessibility schemeIdUri="urn:scte:dash:cc:cea-608:2015"/>
         <Role schemeIdUri="urn:mpeg:dash:role:2011" value="main"/>
         <SegmentTemplate startNumber="0" initialization="$RepresentationID$/init.mp4" duration="6" media="$RepresentationID$/$Number$.m4s"/>
         <ContentProtection schemeIdUri="urn:mpeg:dash:mp4protection:2011" value="cenc"/>
         <ContentProtection schemeIdUri="urn:uuid:edef8ba9-79d6-4ace-a3c8-27dcd51d21ed" value="Widevine: SD"/>
         <ContentProtection schemeIdUri="urn:uuid:9a04f079-9840-4286-ab92-e65be0885f95" value="Playready: SD"/>
         <Representation id="V2200_W" codecs="avc1.4d401f" bandwidth="2384800" mobitv:average-bandwidth="2205862" mobitv:peak-bandwidth="2361517" width="960" height="540" frameRate="60/2" sar="1:1"/>
         <Representation id="V5000_W" codecs="avc1.640020" bandwidth="5120000" mobitv:average-bandwidth="5001378" mobitv:peak-bandwidth="5284876" width="1280" height="720" frameRate="60/2" sar="1:1"/>
         <Representation id="V864_W" codecs="avc1.4d401e" bandwidth="884736" mobitv:average-bandwidth="871321" mobitv:peak-bandwidth="944708" width="640" height="360" frameRate="60/2" sar="1:1"/>
      </AdaptationSet>
      <AdaptationSet contentType="audio" mimeType="audio/mp4" lang="eng" mobitv:qualityLevel="SD" segmentAlignment="true" startWithSAP="1">
         <Role schemeIdUri="urn:mpeg:dash:role:2011" value="main"/>
         <SegmentTemplate startNumber="0" initialization="$RepresentationID$/init.mp4" duration="6" media="$RepresentationID$/$Number$.m4s"/>
         <ContentProtection schemeIdUri="urn:mpeg:dash:mp4protection:2011" value="cenc"/>
         <ContentProtection schemeIdUri="urn:uuid:edef8ba9-79d6-4ace-a3c8-27dcd51d21ed" value="Widevine: SD"/>
         <ContentProtection schemeIdUri="urn:uuid:9a04f079-9840-4286-ab92-e65be0885f95" value="Playready: SD"/>
         <Representation id="A96_ENG" codecs="mp4a.40.5" bandwidth="98304" mobitv:average-bandwidth="101359" mobitv:peak-bandwidth="102064" audioSamplingRate="48000">
            <AudioChannelConfiguration schemeIdUri="urn:mpeg:dash:23003:3:audio_channel_configuration:2011" value="2"/>
         </Representation>
         <InbandEventStream schemeIdUri="urn:mobitv:event" value="id3"/>
      </AdaptationSet>
   </Period>
</MPD>
```

How to start the media right after LICENSE response? First audio and video segments are downloaded before the License call . why to wait till 2nd video segment download ?

**Comment by theodab** [comment]
What happens if you set the `bufferingGoal` a bit lower, like 5 seconds? Or 5.5 seconds?

**Comment by mraavi** [comment]
i tried setting bufferingGoal lower than 6 sec. That didn't help.

**Comment by theodab** [comment]
I'm not sure if I can do much more without actually being able to try out your content. Could you please provide the information necessary to reproduce this (manifest URI, DRM keys, etc)? If anything is confidential, you can send the details to our mailing address: shaka-player-issues@google.com

**Comment by raskri** [comment]
This might help? 
Ignore your minBufferTime="PT2S" by setting:
player.configure('manifest.dash.ignoreMinBufferTime', true);
player.configure('streaming.rebufferingGoal', 0);

**Comment by mraavi** [comment]
@raskri i have tried this. Didn't help.

**Comment by joeyparrish** [comment]
@mraavi, if you can't provide us with a way to reproduce your problem, we will not be able to diagnose the issue.  A copy of your MPD text is not useful without the content.  If you can send an accessible manifest URL to shaka-player-issues@google.com, we can help you further.

Alternately, if the issue is not specific to your content, you could provide us with repro steps using the content in our demo app, which is all public.

**Comment by mraavi** [comment]
@joeyparrish Here is the mpd
https://live-integration-msp-edge.connectcdn.net/mm/dash/live/67812/LIVESERVICE_1234/group=TG_WIFI_HD.mpd
The issue can be reproduced with Shaka demo player as well
https://shaka-player-demo.appspot.com/demo/#audiolang=en-US;textlang=en-US;uilang=en-US;panel=CUSTOM%20CONTENT;build=uncompiled

**Comment by TheModMaker** [comment]
When we start playing, we'll start playing in the middle of a segment, so although we download the first segment, it won't contribute the whole time to the `bufferingGoal`/`rebufferingGoal`.  So after we download the first segment, we may have only buffered 2 seconds.  So if you set the `rebufferingGoal` to 6 seconds, it will require us to always download two segments.  You should set the `rebufferingGoal` to a small number like 2 seconds and the `bufferingGoal` to a larger number like 30 seconds.

By playing your asset and throttling the network, I see us playing after we download the first segment with the default settings.  So try reducing the `rebufferingGoal`.

**Comment by mraavi** [comment]
@TheModMaker I have tried reducing `rebufferingGoal` to 1. It didn't help.
Did the configuration as below
mPlayer.configure('manifest.dash.ignoreMinBufferTime', true);
mPlayer.configure('streaming.rebufferingGoal', 1);

In fact this can be observed with Demo Shaka player and with the mpd i have provided.
I have uploaded a video with screen recording.
https://youtu.be/KnEKneQNigU

**Comment by mraavi** [comment]
Any update on this.

Actually the issue happening with non-live url's as well.
I sent out an email to 'shaka-player-issues@google.com' with subject as 'Faster media start up #2291' and provided non-live url.

**Comment by mraavi** [comment]
@TheModMaker @joeyparrish Please let me know if you need any additional info from my side. Any idea why player needs to download multiple media segments, even though it can start the playback with the first media segment?

**Comment by joeyparrish** [comment]
We'll look into it soon.  Thanks for the URLs, and thanks for your patience.

**Comment by joeyparrish** [comment]
@mraavi, can you confirm what version of Shaka Player you're using?  I don't see that anywhere in your original report.

**Comment by joeyparrish** [comment]
Also, can you please provide your Shaka Player version and the complete configuration you're applying to it?  It's fine to censor license server URLs and such, but we need to see any non-standard config values you're using to completely assess the situation.

**Comment by mraavi** [comment]
We are using Shaka player version v2.5.3.
Configuration applied,
       ```
 mPlayer.configure('manifest.dash.ignoreMinBufferTime', true);
mPlayer.configure('manifest.dash.clockSyncUri', headerUrl);
        mPlayer.configure({
            abr: {
                defaultBandwidthEstimate: 1000000,
                switchInterval: 30
            },
            streaming: {
                bufferBehind: 12,
                bufferingGoal: 30,
                rebufferingGoal: 1
            }
        });
```
**As i stated above in the thread the issue could be reproducible with demo Shaka player as well both with live and non-live urls.

**Comment by mraavi** [comment]
@ismena 
1. You are right about, it takes 2 media segments to finish download to start the media on Chrome PC/MAC browser.
2. But it doesn't not depend on the network condition. On any network it requires 2 Video segments to finish downloading to start Live Playback.
3. Uploaded a video, which shows Live playback starts after 2 video segment download finish - https://youtu.be/kq8ZH0fj_I0. You can see 2 fetched/appended logs for Video and 2 fetched/appended for audio  before video start.
4. Uploaded a Video, which show "Non-Live playback" starts after 1 audio and 1 video segment download finish - https://youtu.be/zHekZqP3x1s
5. Our main concern is, Why would Shaka player waits till the 2nd video segment finish downloading where it can start the playback right after 1st video segment download (like Non-Live)
6. On Samsung TV's (With Tizen OS), Live playback starts after 3rd Video segment download finish. - https://youtu.be/YELEfJi-i54
7. Full log from Samsung TV is sent to 'shaka-player-issues@google.com' with Subject 'Faster media start up #2291'.

**Comment by TheModMaker** [comment]
This is actually expected behavior for live.  The segments in the manifest are only approximations of the times of the segments.  The times in the manifest can be off by half a segment duration.  So we start buffering at the segment behind the current time so we make sure we get something buffered at the playhead.

As for buffering three segments on Tizen, your video is a whole 4 seconds long.  It only appears we "wait" because everything loads so fast.  We check whether we have buffered enough on a timer, so after we have buffered enough, there is a slight delay before we start playing.  There may also be a delay caused by the platform that causes the next segment to complete before you can see anything.

**Comment by mraavi** [comment]
> This is actually expected behavior for live. The segments in the manifest are only approximations of the times of the segments. The times in the manifest can be off by half a segment duration. So we start buffering at the segment behind the current time so we make sure we get something buffered at the playhead.

@TheModMaker you mean to say, For live, Shaka downloads the previous segment before it downloads the current segment? And the playback starts after both previous and current segment download finishes?

**Comment by mraavi** [comment]
> As for buffering three segments on Tizen, your video is a whole 4 seconds long. It only appears we "wait" because everything loads so fast. We check whether we have buffered enough on a timer, so after we have buffered enough, there is a slight delay before we start playing. There may also be a delay caused by the platform that causes the next segment to complete before you can see anything.



> As for buffering three segments on Tizen, your video is a whole 4 seconds long. It only appears we "wait" because everything loads so fast. We check whether we have buffered enough on a timer, so after we have buffered enough, there is a slight delay before we start playing. There may also be a delay caused by the platform that causes the next segment to complete before you can see anything.

In case of Samsung (Tizen), i can confirm that it is (3 segment downloads) not because the segments downloaded too fast. We have done 10's of tests under different network conditions to confirm this. 
Do you think it is a platform issue? Is there a way i can provide you more logs to confirm this?

**Comment by joeyparrish** [comment]
We're adding more logs and more granular configuration for the "download previous segment" behavior.  After that, we'll ask you to try again with the new code and new configuration to see if things have improved.  The new logs will improve our ability to explain to you what's happening in more detail.

**Comment by joeyparrish** [comment]
The latest code in the master branch should both provide better logs and be less conservative about the "download previous segment" behavior to deal with inaccurate manifests.  This behavior is also now configurable.

Can you please try again with the latest code and tell us if things have improved?

If not, can you please attach logs at level v2?  `shaka.log.setLevel(shaka.log.Level.V2);`  (This logging requires a debug or uncompiled build.)

**Comment by mraavi** [comment]
@shaka-bot reopen
We are testing with latest shaka code. will share the logs soon.

**Comment by mraavi** [comment]
Looks like Demo shaka player not available with the main branch.

We have integrated main dev brach in our app and we observe the issue reported still exist,
Logs with "shaka.log.setLevel(shaka.log.Level.V2)" sent to shaka-player-issues@google.com,

**Comment by joeyparrish** [comment]
Can you please be more specific?  I don't know what you mean by "main branch" or "main dev branch".  Can you specify the branch name or release version that you are using from our repo?

I also don't know what you mean by "Demo shaka player not available with the main branch".  Whichever branch you mean, our demo player is available in the `master` branch and all release branches (`v2.5.x`, `v2.4.x`, etc) of our repo.

**Comment by TheModMaker** [comment]
Also those logs aren't with the log level set, or it was severely redacted.  Also be sure to check that you aren't filtering logs in DevTools; make sure you have selected "All Levels" for logs.  V2 logs should be very verbose.  Also you can configure when we fetch the previous segment with:

```js
// Fetch previous segment when within 2 seconds of start of the segment:
player.configure('streaming.inaccurateManifestTolerance', 2);
// Never fetch previous segment:
player.configure('streaming.inaccurateManifestTolerance', 0);
```

Also note the default site is v2.5.9; you can see builds from the `master` branch on our nightly site: <https://nightly-dot-shaka-player-demo.appspot.com/demo/>.

**Comment by mraavi** [comment]
@joeyparrish @TheModMaker 
Thanks for the detail.
We have tested Live demo player (https://nightly-dot-shaka-player-demo.appspot.com/demo/) with **inaccurateManifestTolerance** as 0. Now the playback starts as soon as first segment downloaded.
Also we have integrated, Shaka master branch in our app and tested Live with the configuration (player.configure('streaming.inaccurateManifestTolerance', 0);), Observed the same - Playback starts as soon as first segment and license calls are downloaded.

Observed same on Samsung TV's.

Thanks for the fix. 
When can we expect next version of Shaka player release?

**Comment by joeyparrish** [comment]
We're working on cherry-picks for the next bug fix release (v2.5.10) now.  We will finalize v2.6.0 as soon as #1339 is complete.  This change will not appear in the bug fix release, as the previous behavior was intended and the new behavior might break some applications.

**Comment by joeyparrish** [comment]
Since there doesn't seem to be any more work to do on this issue, I'm going to close it now.  Thanks!
